File: php/sort/quicksort.simple.php
<?php

/**
 * @param array<int> $a
 * @return array<int>
 */
function quicksort(array $a): array
{
    $a_len = count($a);
    if ($a_len < 2) {
        return $a;
    } elseif ($a_len === 2) {
        return ($a[0] > $a[1]) ? [$a[1], $a[0]] : $a;
    }

    $pivot = $a[rand(0, $a_len - 1)];

    $first_part = $equal_part = $second_part = [];
    for ($i = 0; $i < $a_len; $i++) {
        match ($a[$i] <=> $pivot) {
            -1 => $first_part[] = $a[$i],
            0 => $equal_part[] = $a[$i],
            1 => $second_part[] = $a[$i],
        };
    }

    return array_merge(
        quicksort($first_part),
        $equal_part,
        quicksort($second_part),
    );
}

File: go/sort/quicksort.simple.go
package main

import (
	"fmt"
	"math/rand"
)

func quicksort(a []int) []int {
	aLen := len(a)
	if aLen < 2 {
		return a
	} else if aLen == 2 {
		if a[0] > a[1] {
			a[0], a[1] = a[1], a[0]
		}

		return a
	}

	pivot := a[rand.Intn(aLen)]

	var firstPart, equalPart, secondPart []int
	for i := 0; i < aLen; i++ {
		if a[i] < pivot {
			firstPart = append(firstPart, a[i])
		} else if a[i] == pivot {
			equalPart = append(equalPart, a[i])
		} else {
			secondPart = append(secondPart, a[i])
		}
	}

	sortedFirstPart := quicksort(firstPart)
	sortedSecondPart := quicksort(secondPart)

	return append(append(sortedFirstPart, equalPart...), sortedSecondPart...)
}

func main() {
	a := []int{1, 5, 2, 8, 7, 13, 23, 4, 9, 45, 100, 5, 1, 23, 54, 42, 42}

	fmt.Println("Sorted array:", quicksort(a))
}
